# Files used by the scripts in this repo

## Chromosome sizes
The file `hg19.chrom.sizes` was obtained from the database associated with the UCSC Genome Browser using the `fetchChromSizes` script available through the browser. The format is two columns, one for the name of the chromosome, and one for the size of the chromosome. Not all chromosomes named in this file need to be used, but all chromosomes used in the CNV analysis must be present (with their sizes) in this file.

## Bin definitions
The CNV analysis begins with counts of reads in genomic bins. The bins are defined by have roughly equal size, but account for mappability, so the sizes vary slightly. We use files originally generated by scripts provided along with the publication:

Baslan, Timour, et al. "Genome-wide copy number analysis of single cells."
Nature protocols 7.6 (2012): 1024.

We then modified these files to be in 3-column BED format (chrom, start, end). There are files for 5k bins, 20k bins, and 50k bins.

## GC content information
The analysis uses information about GC content of each bin. This is contained in 
