# Data files used by the scripts in this repo

## Chromosome sizes
The file `hg19.chrom.sizes` was obtained from the database associated
with the UCSC Genome Browser using the `fetchChromSizes` script
available through the browser. The format is two columns, one for the
name of the chromosome, and one for the size of the chromosome. Not
all chromosomes named in this file need to be used, but all
chromosomes used in the CNV analysis must be present (with their
sizes) in this file.

## Bin definitions The CNV analysis begins with counts of reads in
genomic bins. The bins are defined by have roughly equal size, but
account for mappability, so the sizes vary slightly. We use files
originally generated by scripts provided along with the publication:
```
Baslan, Timour et al. (2012)
Genome-wide copy number analysis of single cells.
Nature Protocols, 7(6):1024.
```
We then modified these files to be in 3-column BED format (chrom,
start, end). There are files for 5k bins, 20k bins, and 50k bins. The
files are named like `bins_5k_hg19.bed`.

## GC content information
The analysis uses information about GC content of each bin. This
information is contained in files named `bins_5k_hg19_gc.bed` and
these have bedgraph format (chrom, start, end, value). The

## Bad bins
Experience of others who have run similar analyses using a similar
pipeline and the same sets of intervals has shown that certain parts
of the genome attract an unusual amount of reads. For each of the bin
counts (5k, 20k, 50k) a list of "bad bins" are kept in files named
like `bins_5k_hg19_exclude.txt`. The format is simply the line
numbers, in the corresponding bed file, of the bins to exclude from
the CNV analysis.
